.github/workflows/test.yml

```diff
@@ -0,0 +1,77 @@
+name: Tests
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - 'develop'
+      - 'main'
+      - 'release/**'
+      - 'feature/ci'
+  pull_request:
+    branches:
+      - 'develop'
+      - 'main'
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
+        node-version: ['20.x']
+
+    services:
+      postgres:
+        image: postgres:latest
+        env:
+          POSTGRES_DB: edpop
+          POSTGRES_PASSWORD: edpop
+          POSTGRES_USER: edpopuser
+        ports:
+          - 5432:5432
+        options: >-
+          --health-cmd pg_isready
+          --health-interval 10s
+          --health-timeout 5s
+          --health-retries 5
+
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Java 21
+        uses: actions/setup-java@v4
+        with:
+          distribution: 'temurin'
+          java-version: '21'
+      - name: Download and run Blazegraph
+        run: |
+          wget https://github.com/blazegraph/database/releases/download/BLAZEGRAPH_2_1_6_RC/blazegraph.jar
+          java -jar blazegraph.jar &
+          sleep 5
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+          cache: 'pip'
+      - name: Install Python packages
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r backend/requirements.txt
+          pip install -r backend/requirements-test.txt
+      - name: Test with pytest
+        run: |
+          cd backend && pytest
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v4
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: 'npm'
+          cache-dependency-path: './frontend/package-lock.json'
+      - name: Install frontend dependencies
+        run: |
+          cd frontend
+          npm ci
+      - name: Run frontend tests
+        run: |
+          cd frontend
+          npm run test
```

Comment:
I think this is completely independent of the backend installation + tests in previous steps, so you could split these steps into a separate job (currently, everything is a single "build" job). That way, the frontend and backend tests can be run in parallel, so the action will run faster.

Reply:
OK!